### PR TITLE
Wildcard make

### PR DESCRIPTION
**Pull request by matatk**
This PR changes the behaviour of the Makefile to generate test binaries for each file matching `*.test.c`.

It also includes basic installation instructions in the README (I needed somewhere to write them down anyway…)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,3 +4,21 @@ TESTickle
 [![Build Status](https://travis-ci.org/MarkWithall/TESTickle.svg?branch=master)](https://travis-ci.org/MarkWithall/TESTickle)
 
 A unit testing library for C.
+
+Installation
+------------
+
+### Windows
+
+Download and install [Strawberry Perl](http://strawberryperl.com); this comes with `bcc` and `gmake` too. You will need to install the required CPAN libraries with: `cpan Template`
```
Comment (MarkWithall) [inline README.md]: Should this be `gcc` rather than `bcc`?
Also, does the `cpan` command require the `-i` switch for installing?

**Comment by MarkWithall** [comment]
Thanks for this.  The Makefile improvements look excellent.  Just a couple of minor typos in the README, I think.

**Comment by matatk** [comment]
I'm glad that I had StackOverflow as well as the manual to help with the changes :-).  Fixed the typo; ta!

Re CPAN, my interpretation of [the cpan man page](http://www.manpagez.com/man/1/cpan/) is that you don't need to specify `-i` (though you do if you're also specifying `-f` like in your Travis YAMLs).

The `cpanm` command for Perlbrew is not a typo, though it looks like it could be one :-).

**Comment by MarkWithall** [comment]
Yes, the command `cpan Template` works on mine.

**Comment by MarkWithall** [comment]
Alas, the Makefile fails in Windows, using `dmake` with following errors:

```
dmake:  makefile:  line 34:  Warning: -- More than one prerequisite
for %-target. Use :| ruleop or indirect prerequisites.
dmake:  makefile:  line 40:  Warning: -- More than one prerequisite
for %-target. Use :| ruleop or indirect prerequisites.
```

I shall have a look into it and see if I can find a fix that works.

**Comment by MarkWithall** [comment]
Changing the `:` on lines 34 and 40 to `:|` makes the errors go away.

However, the list of output files doesn't appear to be being generated by the new `all: ...patsubst...` part.  And if I run the target file explicitly, I get:

```
dmake:  Warning: -- Ambiguous inference chains for target 'bin\testickle.exe'
dmake:   1. bin\testickle.exe(bin\%.exe) --> obj\testickle.o(obj\%.o) --> TESTickle.h(TESTickle.h)
dmake:   2. bin\testickle.exe(bin\%.exe) --> obj\testickle.o(obj\%.o) --> testickle.test.c(%.test.c)
dmake:   3. bin\testickle.exe(bin\%.exe) --> testickle_runner.c(%_runner.c) --> testickle.test.c(%.test.c)
dmake:  Warning: -- First matching rule is chosen.
gcc -fms-extensions -std=c99 -Wall -Wextra -Werror -g -o obj\testickle.o -c TESTickle.h
gcc -fms-extensions -std=c99 -Wall -Wextra -Werror -g -o bin\testickle.exe obj\testickle.o
obj\testickle.o: file not recognized: File format not recognized
collect2.exe: error: ld returned 1 exit status
dmake:  Error code 129, while making 'bin\testickle.exe'
```

**Comment by matatk** [comment]
Shame... though what's `dmake`?  We were using `gmake` that comes with Strawberry Perl for previous use of TESTickle on Windows.  I have not actually tested this with `gmake` though (sorry) -- but I have tried to use Make-internal constructs (rather than shell stuff), so had hoped it would work with `gmake` because I think that's essentially standard GNU Make. [Side note: `/usr/bin/make` on the latest OS X release is GNU make.]

If I find the cycles to test this on Windows with `gmake` too I'll report back.

**Comment by MarkWithall** [comment]
`dmake` also comes with Strawberry.  I hadn't noticed `gmake` was there too.  Alas, it also doesn't work with different errors.

```
perl test_generator.pl testickle.test.c > testickle_runner.c
gcc -fms-extensions -std=c99 -Wall -Wextra -Werror -g -o bin/testickle.exe testickle_runner.c obj/testickle.o
obj/testickle.o: file not recognized: File format not recognized
collect2.exe: error: ld returned 1 exit status
gmake: *** [bin/testickle.exe] Error 1
rm testickle_runner.c
```

**Comment by MarkWithall** [comment]
Looks like it's skipping a step.

**Comment by MarkWithall** [comment]
Alas, it is still missing the .o build step.

**Comment by matatk** [comment]
Just for reference, here's the output I get when I have copied testickle.test.c to moo.test.c in the same directory, and run it on OS X...

```
➜  TESTickle make
src files: moo.test.c testickle.test.c
bin_names: moo testickle
bin_paths: bin/moo bin/testickle
obj_names: moo.o testickle.o
obj_paths: obj/moo.o obj/testickle.o

mkdir -p obj bin
perl test_generator.pl moo.test.c > moo_runner.c
gcc -fms-extensions -std=c99 -Wall -Wextra -Werror -g -o obj/moo.o -c moo.test.c
gcc -fms-extensions -std=c99 -Wall -Wextra -Werror -g -o bin/moo moo_runner.c obj/moo.o
perl test_generator.pl testickle.test.c > testickle_runner.c
gcc -fms-extensions -std=c99 -Wall -Wextra -Werror -g -o obj/testickle.o -c testickle.test.c
gcc -fms-extensions -std=c99 -Wall -Wextra -Werror -g -o bin/testickle testickle_runner.c obj/testickle.o
rm moo_runner.c obj/moo.o testickle_runner.c obj/testickle.o
```

**Comment by matatk** [comment]
I think I'm doing something wrong with the three-clause target specifiers, as in [this similar example on StackOverflow](http://stackoverflow.com/a/4972155) it has the first clause being a list. Given that I can't remember if I have/haven't yet, I should probably try to put `$(OBJ_PATHS)` as the first part of [line 49](https://github.com/MarkWithall/TESTickle/pull/4/files#diff-b67911656ef5d18c4ae36cb6741b7965R49).

I should also perhaps try reading [the fine manual](https://www.gnu.org/software/make/manual/html_node/Static-Usage.html#Static-Usage) in more depth :-).

**Comment by MarkWithall** [comment]
Just seems to need `\\` for the windows paths `OBJECTS` and `BINARIES` to make `gmake clean` work.  The rest seems to be fine.

**Comment by matatk** [comment]
Great. I have made the `\\` change, but, before I push: what would you like me to do with the "banner" (debugging info) stuff?  I could take it out, now we are confident that it works, or leave it in for future reassurance.

**Comment by MarkWithall** [comment]
Thanks.  I think we should take the banner out for now.

**Comment by matatk** [comment]
P.S. I also wonder if it's worth me adding comments to point out that there is a difference between `OBJECTS` and `BINARIES` defined, in the conditional section, for use by the `clean` target (which will pick up any existing objects or binaries) and those computed based on the `*.test.c` files in the directory when the Makefile is run?  I think it's important to have the two separated out, because then `clean` really does a full clean, but it may not be terribly clear from the present Makefile code as to why there are two different techniques used to do (at first glance) the same thing.

**Comment by MarkWithall** [comment]
Sounds like a good idea.

**Comment by matatk** [comment]
Done... I'm still not sure if I think that the platform-specific stuff should be after the dynamic stuff -- on the one hand, it feels more appropriate for the platform-specific config to be closer to the top of the Makefile because it's more fundamental platform-related stuff, but on the other hand I wonder if the comments around the dynamic construction of targets based on `*.test.c` would make more sense if _they_ were higher up.

**Comment by MarkWithall** [comment]
Yes.  That works on mine now.  I've added issues to address the location of the test specific make code by putting it into a separate makefile and add back support for dmake (probably via a second makefile that uses the include).

**Pull request merged**